// mojito-create/index.js
function run(args, cb) {

}

exports.usage = [
    'mojito create {type} [subtype or source directory] {name} [options]',
    '  - type: "app", "mojit", or "custom".',
    '  - archetype: optional template. Possible values are default, full, simple',
    '      "app" types also have a "hybrid" archetype which creates an app and a',
    '      mojit with common configurations for use with hybrid app.',
    '      If the type is "custom" then this is the path to your own archetype',
    '      directory.',
    '  - name: name of the app or mojit to create',
    '',
    'Example: mojito create app Foo',
    "  (creates directory 'Foo' containing new Mojito application)",
    '',
    'Example: mojito create mojit Bar',
    '  (creates directory "Bar" containing new Mojit)',
    '',
    'Example: mojito create app hybrid Baz',
    '  (creates directory "Baz" containing new hybrid app and mojit)',
    '',
    'OPTIONS: ',
    '  --port [number]  Specifies default port for your Mojito app to run on.',
    '  -p [number]      Short for --port',
    '  --force          Forced mojit creation even outside a Mojito app.',
    '  -f               Short for --force',
    '  -keyval [string] key value pairs to pass to a custom archetype template',
    '                   a key/value is separated by colons, key/value pairs by',
    '                   commas: "key1:val1,key2:val2',
    '  -k [string]      Short for --keyval'].join("\n");

exports.options = [{
    shortName: 'f',
    longName: 'force',
    hasValue: false
}, {
    shortName: 'k',
    longName: 'keyval',
    hasValue: true
}, {
    shortName: 'p',
    longName: 'port',
    hasValue: true
}];

exports.run = run;
